### ISSUE TITLE
Incompatibility with CometD library (>=v3.1.1)

### ISSUE THREAD

**Issue by vanjabucic** [faye/faye#497]
**Problem**
We are seeing our clients reconnect once a minute (on second /meta/connect exchange) even though it is not obvious why.
There are packets being sent every 15 seconds which should be plenty enough.
From what I can see, the Faye client is not waiting long enough for /meta/connect reply that is sent by CometD server every 30 seconds.
Oddly enough, if we increase the publish rate to every 5 seconds the client will correctly wait 30 seconds and not timeout.

**Environment**: Debian9, ruby: 2.3.4p301, latest faye bundle, cometd-version=3.1.1, jetty-version=9.2.22.v20170606
Tests are done on a local network, without proxies and firewalls.

**Log**
```
I, [2018-06-07T12:56:04.374059 #18354]  INFO -- : [Faye::Client] New client created for "http://10.0.0.198:8080/endpoint/v1"
I, [2018-06-07T12:56:04.379172 #18354]  INFO -- : [Faye::Client] Initiating handshake with "http://10.0.0.198:8080/endpoint/v1"
D, [2018-06-07T12:56:04.394345 #18354] DEBUG -- : [Faye::Dispatcher] Selected "long-polling" transport for http://10.0.0.198:8080/endpoint/v1
D, [2018-06-07T12:56:06.275612 #18354] DEBUG -- : [Faye::Transport::Http] Client nil sending message to http://10.0.0.198:8080/endpoint/v1 via "long-polling": {"channel":"/meta/handshake","version":"1.0","supportedConnectionTypes":["in-process","websocket","long-polling"],"id":"1","ext":{"com.company.tagevent":{..}}}
D, [2018-06-07T12:56:06.405380 #18354] DEBUG -- : [Faye::Transport::Http] Client nil received from http://10.0.0.198:8080/endpoint/v1 via "long-polling": [{"minimumVersion":"1.0","clientId":"7tlsqflk2z6bkp4v14euu58nrh7g5","supportedConnectionTypes":["websocket","long-polling","callback-polling"],"advice":{"interval":0,"timeout":30000,"reconnect":"retry"},"channel":"/meta/handshake","id":"1","version":"1.0","successful":true}]

D, [2018-06-07T12:56:06.406241 #18354] DEBUG -- : [Faye::Dispatcher] Selected "websocket" transport for http://10.0.0.198:8080/endpoint/v1
I, [2018-06-07T12:56:06.406351 #18354]  INFO -- : [Faye::Client] Handshake successful: "7tlsqflk2z6bkp4v14euu58nrh7g5"
I, [2018-06-07T12:56:06.406458 #18354]  INFO -- : [Faye::Client] Calling deferred actions for "7tlsqflk2z6bkp4v14euu58nrh7g5"
D, [2018-06-07T12:56:06.407008 #18354] DEBUG -- : [Faye::Transport::WebSocket] Client "7tlsqflk2z6bkp4v14euu58nrh7g5" sending message to http://10.0.0.198:8080/endpoint/v1 via "websocket": {"channel":"/service/tagevent/registerDataSource","data":{..},"clientId":"7tlsqflk2z6bkp4v14euu58nrh7g5","id":"2"}


I, [2018-06-07T12:56:06.407352 #18354]  INFO -- : [Faye::Client] Initiating connection for "7tlsqflk2z6bkp4v14euu58nrh7g5"
D, [2018-06-07T12:56:06.407692 #18354] DEBUG ** :*[Faye::Transport::WebSocket] Client "7tlsqflk2z6bkp4v14euu58nrh7g5" sending message to http://10.0.0.198:8080/endpoint/v1 via "websocket": {"channel":"/meta/connect","clientId":"7tlsqflk2z6bkp4v14euu58nrh7g5","connectionType":"websocket","id":"3"}


D, [2018-06-07T12:56:06.410587 #18354] DEBUG -- : [Faye::Transport::WebSocket] Client "7tlsqflk2z6bkp4v14euu58nrh7g5" sending message to http://10.0.0.198:8080/endpoint/v1 via "websocket": {"channel":"/meta/subscribe","clientId":"7tlsqflk2z6bkp4v14euu58nrh7g5","subscription":"/tagevent/subscribe/feedback","id":"4"}
D, [2018-06-07T12:56:06.410873 #18354] DEBUG -- : [Faye::Transport::WebSocket] Client "7tlsqflk2z6bkp4v14euu58nrh7g5" received from http://10.0.0.198:8080/endpoint/v1 via "websocket": [{"data":{"status":"OK"},"channel":"/service/tagevent/registerDataSource"}]

D, [2018-06-07T12:56:06.413005 #18354] DEBUG ** :*[Faye::Transport::WebSocket] Client "7tlsqflk2z6bkp4v14euu58nrh7g5" received from http://10.0.0.198:8080/endpoint/v1 via "websocket": [{"advice":{"interval":0,"timeout":30000,"reconnect":"retry"},"channel":"/meta/connect","id":"3","successful":true}]
I, [2018-06-07T12:56:06.413315 #18354]  INFO -- : [Faye::Client] Closed connection for "7tlsqflk2z6bkp4v14euu58nrh7g5"
I, [2018-06-07T12:56:06.413445 #18354]  INFO -- : [Faye::Client] Calling deferred actions for "7tlsqflk2z6bkp4v14euu58nrh7g5"
I, [2018-06-07T12:56:06.413523 #18354]  INFO -- : [Faye::Client] Initiating connection for "7tlsqflk2z6bkp4v14euu58nrh7g5"
D, [2018-06-07T12:56:06.413860 #18354] DEBUG ** :*[Faye::Transport::WebSocket] Client "7tlsqflk2z6bkp4v14euu58nrh7g5" sending message to http://10.0.0.198:8080/endpoint/v1 via "websocket": {"channel":"/meta/connect","clientId":"7tlsqflk2z6bkp4v14euu58nrh7g5","connectionType":"websocket","id":"5"}

D, [2018-06-07T12:56:06.414830 #18354] DEBUG -- : [Faye::Transport::WebSocket] Client "7tlsqflk2z6bkp4v14euu58nrh7g5" received from http://10.0.0.198:8080/endpoint/v1 via "websocket": [{"channel":"/meta/subscribe","id":"4","subscription":"/tagevent/subscribe/feedback","successful":true}]
D, [2018-06-07T12:56:06.415958 #18354] DEBUG -- : [Faye::Transport::WebSocket] Client "7tlsqflk2z6bkp4v14euu58nrh7g5" sending message to http://10.0.0.198:8080/endpoint/v1 via "websocket": {"channel":"/service/tagevent/publishEvents","data":{..},"clientId":"7tlsqflk2z6bkp4v14euu58nrh7g5","id":"6"}
D, [2018-06-07T12:56:06.417335 #18354] DEBUG -- : [Faye::Transport::WebSocket] Client "7tlsqflk2z6bkp4v14euu58nrh7g5" received from http://10.0.0.198:8080/endpoint/v1 via "websocket": [{"data":{"status":"OK"},"channel":"/service/tagevent/publishEvents"}]
D, [2018-06-07T12:56:06.417938 #18354] DEBUG -- : [Faye::Transport::WebSocket] Client "7tlsqflk2z6bkp4v14euu58nrh7g5" received from http://10.0.0.198:8080/endpoint/v1 via "websocket": [{"channel":"/service/tagevent/publishEvents","id":"6","successful":true}]


D, [2018-06-07T12:56:08.421337 #18354] DEBUG -- : [Faye::Transport::WebSocket] Client "7tlsqflk2z6bkp4v14euu58nrh7g5" sending message to http://10.0.0.198:8080/endpoint/v1 via "websocket": {"channel":"/service/tagevent/publishEvents","data":{..},"clientId":"7tlsqflk2z6bkp4v14euu58nrh7g5","id":"7"}
D, [2018-06-07T12:56:08.423248 #18354] DEBUG -- : [Faye::Transport::WebSocket] Client "7tlsqflk2z6bkp4v14euu58nrh7g5" received from http://10.0.0.198:8080/endpoint/v1 via "websocket": [{"data":{"status":"OK"},"channel":"/service/tagevent/publishEvents"}]
D, [2018-06-07T12:56:08.423895 #18354] DEBUG -- : [Faye::Transport::WebSocket] Client "7tlsqflk2z6bkp4v14euu58nrh7g5" received from http://10.0.0.198:8080/endpoint/v1 via "websocket": [{"channel":"/service/tagevent/publishEvents","id":"7","successful":true}]


D, [2018-06-07T12:56:28.444482 #18354] DEBUG -- : [Faye::Transport::WebSocket] Client "7tlsqflk2z6bkp4v14euu58nrh7g5" sending message to http://10.0.0.198:8080/endpoint/v1 via "websocket": {"channel":"/service/tagevent/publishEvents","data":{..},"clientId":"7tlsqflk2z6bkp4v14euu58nrh7g5","id":"8"}
D, [2018-06-07T12:56:28.446464 #18354] DEBUG -- : [Faye::Transport::WebSocket] Client "7tlsqflk2z6bkp4v14euu58nrh7g5" received from http://10.0.0.198:8080/endpoint/v1 via "websocket": [{"data":{"status":"OK"},"channel":"/service/tagevent/publishEvents"}]
D, [2018-06-07T12:56:28.447008 #18354] DEBUG -- : [Faye::Transport::WebSocket] Client "7tlsqflk2z6bkp4v14euu58nrh7g5" received from http://10.0.0.198:8080/endpoint/v1 via "websocket": [{"channel":"/service/tagevent/publishEvents","id":"8","successful":true}]

                                            [30 sec since last connect]
D, [2018-06-07T12:56:36.415291 #18354] DEBUG ** :*[Faye::Transport::WebSocket] Client "7tlsqflk2z6bkp4v14euu58nrh7g5" received from http://10.0.0.198:8080/endpoint/v1 via "websocket": [{"channel":"/meta/connect","id":"5","successful":true}]
I, [2018-06-07T12:56:36.415731 #18354]  INFO -- : [Faye::Client] Closed connection for "7tlsqflk2z6bkp4v14euu58nrh7g5"
I, [2018-06-07T12:56:36.415837 #18354]  INFO -- : [Faye::Client] Calling deferred actions for "7tlsqflk2z6bkp4v14euu58nrh7g5"
I, [2018-06-07T12:56:36.415945 #18354]  INFO -- : [Faye::Client] Initiating connection for "7tlsqflk2z6bkp4v14euu58nrh7g5"
D, [2018-06-07T12:56:36.416296 #18354] DEBUG ** :*[Faye::Transport::WebSocket] Client "7tlsqflk2z6bkp4v14euu58nrh7g5" sending message to http://10.0.0.198:8080/endpoint/v1 via "websocket": {"channel":"/meta/connect","clientId":"7tlsqflk2z6bkp4v14euu58nrh7g5","connectionType":"websocket","id":"9"}
                                            [next one expected @ 12:57:06.4]

I, [2018-06-07T12:56:48.462799 #18354]  INFO -- : [Faye::Client] Calling deferred actions for "7tlsqflk2z6bkp4v14euu58nrh7g5"
D, [2018-06-07T12:56:48.463493 #18354] DEBUG -- : [Faye::Transport::WebSocket] Client "7tlsqflk2z6bkp4v14euu58nrh7g5" sending message to http://10.0.0.198:8080/endpoint/v1 via "websocket": {"channel":"/service/tagevent/publishEvents","data":{..},"clientId":"7tlsqflk2z6bkp4v14euu58nrh7g5","id":"a"}
D, [2018-06-07T12:56:48.465144 #18354] DEBUG -- : [Faye::Transport::WebSocket] Client "7tlsqflk2z6bkp4v14euu58nrh7g5" received from http://10.0.0.198:8080/endpoint/v1 via "websocket": [{"data":{"status":"OK"},"channel":"/service/tagevent/publishEvents"}]
D, [2018-06-07T12:56:48.465823 #18354] DEBUG -- : [Faye::Transport::WebSocket] Client "7tlsqflk2z6bkp4v14euu58nrh7g5" received from http://10.0.0.198:8080/endpoint/v1 via "websocket": [{"channel":"/service/tagevent/publishEvents","id":"a","successful":true}]

                                            [premature connect timeout?]
D, [2018-06-07T12:57:01.569915 #18354] DEBUG ** :*[Faye::Transport::WebSocket] Client "7tlsqflk2z6bkp4v14euu58nrh7g5" failed to send to http://10.0.0.198:8080/endpoint/v1 via "websocket": [{"channel":"/meta/connect","clientId":"7tlsqflk2z6bkp4v14euu58nrh7g5","connectionType":"websocket","id":"9"}]
D, [2018-06-07T12:57:01.570197 #18354] DEBUG -- : [Faye::Transport::WebSocket] Client "7tlsqflk2z6bkp4v14euu58nrh7g5" sending message to http://10.0.0.198:8080/endpoint/v1 via "websocket": {"channel":"/meta/connect","clientId":"7tlsqflk2z6bkp4v14euu58nrh7g5","connectionType":"websocket","id":"9"}
I, [2018-06-07T12:57:01.570863 #18354]  INFO -- : [Faye::Client] Disconnecting "7tlsqflk2z6bkp4v14euu58nrh7g5"
D, [2018-06-07T12:57:01.571225 #18354] DEBUG -- : [Faye::Transport::WebSocket] Client "7tlsqflk2z6bkp4v14euu58nrh7g5" sending message to http://10.0.0.198:8080/endpoint/v1 via "websocket": {"channel":"/meta/disconnect","clientId":"7tlsqflk2z6bkp4v14euu58nrh7g5","id":"b"}
I, [2018-06-07T12:57:01.571310 #18354]  INFO -- : [Faye::Client] Clearing channel listeners for "7tlsqflk2z6bkp4v14euu58nrh7g5"
```

**Comment by vanjabucic** [comment]
**Update on the issue**
It is caused by Faye 'ping' operation. Had to search high and low as it is not part of the trace (not reporting the ping action or the bytes going over the wire).

Using CometD library 3.1.1 it looks like it works but the ping operation is inadvertently  scheduling the session for a sweep, and within 2-3 cycles it is gone.
(empty message fall-through sets scheduleExpiration to true)

Using CometD library 3.1.1+ it flat out fails as the ping is rejected and explicitly throws IOException
[https://github.com/cometd/cometd/blob/846ceb4d627252699a4dcaa44db3766793db428a/cometd-java/cometd-java-websocket/cometd-java-websocket-common-server/src/main/java/org/cometd/websocket/server/common/AbstractWebSocketTransport.java] processMessages()
```java
private void processMessages(S wsSession, ServerMessage.Mutable[] messages) throws IOException {
            if (messages.length == 0) {
                throw new IOException();
            }
...
}
```
Fairly critical issue...

How can we disable/modify the ping behavior?

**Comment by vanjabucic** [comment]
I asked Simone Bordet about empty frames, his answer is here:
[cometd-users topic](https://groups.google.com/forum/#!topic/cometd-users/OLOw0KkgkFo)

Short answer: no empty frames will be accepted, there are other ways to do 'ping'.

Quick fix for others looking for it is to use Ruby magic and rewrite ping method to do nothing:
```ruby
# After Faye is fully loaded:
# Monkeypatch WebSocket to be cometD compatible: disable ping
  Faye::Transport::WebSocket.class_eval do
    private
    def ping
      # return unless @socket and @socket.ready_state == 1
      # @socket.send('[]')
      # add_timeout(:ping, @dispatcher.timeout / 2) { ping }
      puts 'Faye::Transport::WebSocket [ping disabled]'
    end
  end
```

Regarding a long term resolution, @jcoglan can you read Simone's take on it and chime in with a desired direction of a solution?

**Comment by vanjabucic** [comment]
bump..
Any input would be appreciated @jcoglan

**Comment by jcoglan** [comment]
Hi, apologies for the delay on this. I have a different interpretation of the spec to the CometD authors in that, since the server essentially iterates over the messages and generates a response independently for each one (it's a `map` operation), sending an empty list is entirely valid and I'm not sure why their server special-cases it when the subsequent code would handle an empty list just fine.

Still, the fact of the matter is their server won't accept this so I'm removing the functionality -- see https://github.com/faye/faye/commit/45a1731cc2ce7ea8e9b4dc3557266e424cd637f6. If users have issues with proxies timing out their connections, I would suggest lowering the `timeout` on the server to a lower value, so that the client sends `/meta/connect` messages more often.

Thanks for reporting this and following up with the CometD team :)

**Comment by jcoglan** [comment]
Just letting you know that the fix for this issue has finally been released in version 1.3.0.
